### ISSUE TITLE
Milestone: SBOM Generation /  Tooling Best Practices Created

### ISSUE THREAD

**Issue by bensternthal** [openjs-foundation/security-collab-space#18]
```[tasklist]
### Tasks
- [ ] Publish V1 of SBOM Guide For OpenJS Projects
```

**Comment by bensternthal** [comment]
## Links of Interest
* https://cyclonedx.org/

**Comment by ljharb** [comment]
For OpenJS, I think the goal should possibly be to produce a guide, containing trivial/simple instructions for a JS package maintainer to produce valid SBOM(s) that meet the needs of end users.

Whether the SBOM tooling ecosystem is ready for this guide or not remains to be established.

**Comment by ctcpip** [comment]
as a starting point, it probably doesn't get more trivial than `cdxgen -o sbom.json`

**Comment by ljharb** [comment]
Here's my bare initial SBOM doc, full of TODOs: https://hackmd.io/ujVlBEjqQCSvxfspqd3gcw I'd love feedback.

**Comment by ctcpip** [comment]
looks good for the most part.  finished reviewing.  as there are a number of TODOs, let me know when this reaches the _next stage/draft version_

**Comment by ctcpip** [comment]
while it's true there is not broad agreement re: SBOMs, this document should at least guide folks to how to create SBOMs, automation, etc.

while there are a number of formats and such, I think it's reasonable to take a cue from the OpenSSF SBOM Everywhere SIG:

> For the purposes of establishing ubiquity to ensure sustainability for SBOM related tooling, and future solutions for consumption, “supported” formats must be defined. At this time there are two supported formats that will be in scope for the purposes of this group: [CycloneDX](https://cyclonedx.org/) and [SPDX](https://spdx.dev/).

Those two are the most widely adopted and supported, and have broad support for related capabilities like VDR, VEX, etc.

**Comment by ljharb** [comment]
I agree, but for JS devs, any such tool would need to be purely npm-installable. Prior to npm itself supporting SBOMS (which is an ongoing RFC), what would you suggest?

**Comment by ctcpip** [comment]
https://github.com/openjs-foundation/security-collab-space/issues/18#issuecomment-1648380803

**Comment by christian-bromann** [comment]
I am happy to have WebdriverIO be a guinea pig project adapting the recommended way to produce SBOMS. One questions I had for this is: how should they be stored? Attached to a release? If so, in case a project is running a mono repo and a release contains updates to several NPM packages, do we create different SBOMS and attach it with the name of the package as filename?

**Comment by ctcpip** [comment]
simplest thing is to have the sbom generated as part of an existing (or new) CI action, which generates the SBOM and commits it.  that also takes care of it being attached to the release (though projects may choose to package it differently).

for a monorepo, the simplest and easiest (probably) thing is to just let the tool generate the SBOM for the entire monorepo, which is possible via recursion flags.  so it's just one SBOM for the entire thing.  I can provide an example if that's helpful.

if you wanted to, you could create individual SBOMs for each component of the monorepo, but more likely than not, consumers will prefer to have the one SBOM for everything.

**Comment by ljharb** [comment]
Is there a convention yet for eventually being able to include all of the SBOM types, not just source?

**Comment by ctcpip** [comment]
> Is there a convention yet for eventually being able to include all of the SBOM types, not just source?

I'm not sure what you mean? specifically what is meant by 1. SBOM types and 2. source ?

**Comment by ljharb** [comment]
There’s like 7 types of SBOM - the source, but also the environment, the dev tools used to build, etc. only producing one type without a plan for the rest seems shortsighted.

**Comment by ctcpip** [comment]
ah. IME, referring to SBOM has always referred to the artifact source, e.g. the content of a repo.  for provenance and attestation of how an artifact was built then you're looking at SLSA, and you get [SLSA Level 3 via GH Actions](https://github.com/slsa-framework/slsa-github-generator)

**Comment by mhdawson** [comment]
> Here's my bare initial SBOM doc, full of TODOs: https://hackmd.io/ujVlBEjqQCSvxfspqd3gcw I'd love feedback.

I get a permission denied on ^^^

**Comment by ctcpip** [comment]
@mhdawson I think you just need to be signed in

**Comment by ctcpip** [comment]
`cdxgen` I referenced above works for all (supported) languages.  there is also https://github.com/CycloneDX/cyclonedx-node-npm specifically for npm packages.  I have not tested to see if they produce the exact same output, but I should hope they would.

and FWIW:  GH workflow example which uses https://github.com/CycloneDX/gh-node-module-generatebom:

https://github.com/afterburner-js/afterburner-js/blob/sbom/.github/workflows/sbom.yml

you can, of course, also just add a script to your project to generate the SBOM and run it in a workflow, but then you also have the added dev dependency for `cyclonedx-node-npm` or `cdxgen`
